specification/under_development/profiles/Base_Profile.md

```diff
@@ -0,0 +1,522 @@
+
+Both the `"entry_point"` attribute and the `"output_labeling_schema"` attribute
+can only be attached to a function definition; they are invalid on a function
+that is declared but not defined. For the Base Profile, this implies that they
+can occur only in one place.
+
+Within the restrictions imposed by the Base Profile, the number of qubits that
+are needed to execute the quantum program must be known at compile time. This
+number is captured in the form of the `"required_num_qubits"` attribute attached
+to the entry point. The value of the attribute must be the string representation
+of a non-negative 64-bit integer constant.
+
+Similarly, the number of measurement results that need to be stored when
+executing the entry point function is captured by the `"required_num_results"`
+attribute. Since qubits cannot be used after measurement, in the case of the
+Base Profile, this value is usually equal to the number of measurement results
+in the program output.
+
+Beyond the entry point specific requirements related to attributes, custom
+attributes may optionally be attached to any of the declared functions. The
+`irreversible` attribute in particular impacts how the program logic in the
+entry point is structured, as described in the section about the [entry point
+definition](#function-body). Furthermore, the following [LLVM
+attributes](https://llvm.org/docs/LangRef.html#function-attributes) may be used
+according to their intended purpose on function declarations and call sites:
+`inlinehint`, `nofree`, `norecurse`, `readnone`, `readonly`, `writeonly`, and
+`argmemonly`.
+
+## Module Flags Metadata
+
+The following [module
+flags](https://llvm.org/docs/LangRef.html#module-flags-metadata) must be added
+to the QIR bitcode:
+
+- a flag with the string identifier `"qir_major_version"` that contains a
+  constant value of type `i32`
+- a flag with the string identifier `"qir_minor_version"` that contains a
+  constant value of type `i32`
+- a flag with the string identifier `"dynamic_qubit_management"` that contains a
+  constant `true` or `false` value of type `i1`
+- a flag with the string identifier `"dynamic_result_management"` that contains
+  a constant `true` or `false` value of type `i1`
+
+These flags are attached as `llvm.module.flags` metadata to the module. They can
+be queried using the standard LLVM tools and follow the LLVM specification in
+behavior and purpose. Since module flags impact whether different modules can be
+merged and how, additional module flags may be added to the bitcode only if
+their behavior is set to `Warning`, `Append`, `AppendUnique`, `Max`, or `Min`.
+It is at the discretion of the maintainers for various components in the QIR
+stack to discard module flags that are not explicitly required or listed as
+optional flags in the QIR specification.
+
+### Specification Version
+
+The required flags `"qir_major_version"` and `"qir_minor_version"` identify the
```

Comment:
A significant question here - is this major version the major version of the _base profile_, or of QIR as a whole?

If the former, we should be all set thanks to the `qir_profile` entrypoint attribute, but it might be worth calling that relationship out here (`identify the...version of the Base Profile specification to which the QIR bitcode adheres`). Also - if there are two different entrypoints meeting two different profiles (because a new profile may specify a different entrypoint attribute), which minor and major version should be specified?

That is, should the major/minor version actually correspond to the version of _profile_ and thus be an attribute on the entrypoint rather than the module?

If the latter, what if we want to make breaking changes to the Base Profile after other profiles (i.e. major versions) have been published? Does that become a completely new profile with its own major version? That is, would BaseV2 possibly be major version 5, which might be much more restrictive than MoreThanBaseV1 which happens to be major version 4?

Reply:
Thinking a bit more about the discussion we had offline regarding whether or not to version profile specifications independently, I have a couple of thoughts that I wanted to capture:

- It is worth thinking carefully about what profile specific restrictions could make sense that are *not* at the function level. There are actually very few, I believe, currently in the Base Profile spec; 

  - how to label an entry point and 
  - requirements for specific module flags
  - requirements for qis/rt functions that are called by an entry point
  
  are actually the only two that come to mind - the rest are specific rules that either directly apply to the function definition/implementation only (like specific block structures, what runtime functions can be used, what llvm instructions can be used etc.), or what requirements module level constructs need to satisfy to be usable in the entry point (like QIS functions need to return void). 

- The following should ultimately move into the full QIR spec (i.e. it is not really profile specific, but it is not currently in the full spec and I needed it to spec the profile): 

  - Only the first sentence in the section on output recording is specific to the Base Profile - everything else really should go into the full spec.
  - The whole section on attributes is not really specific to the Base Profile, and can pretty much be moved into the full spec. The only Base Profile specific requirements may be that `required_num_qubits` and `required_num_results` should be optional in general and are required for a base profile compliant entry point.
  - Pretty much the same goes for the Module Flags section, where the only Base Profile specific requirement is that the value of the `dynamic_qubit_management` and `dynamic_result_management` must be false. 
  
Given that, my thinking is that a QIR profile ultimately defines rules for the entry point definition, and everything that is used as part of that, but that the rest (like how to identify an entry point) is captured as part of the full specification, and I would keep versioning in sync (as discussed in the SC meeting). I don't think this would cause issues for backends to clearly define what they support; the spec version number defines how the program logic is defined within a bitcode file (that may contain more than only what is necessary/required for a base profile compliant function), and the attribute on the specific function to invoke defines what profile it meets.

Reply:
Okay, I see what you're saying, and I agree with that. I see how it makes sense to annotate the entrypoint rather than the module - although if the module can only have one entrypoint (per whatever is specified by that version of QIR), I don't know that I see much of a difference.

Given that the current paradox here is that the compiler must know the profile in order to find the entrypoint, but it must know the entrypoint to know the profile - if the entrypoint is defined as part of _QIR_ and not the profile, then the paradox is solved.

Further, it makes sense to me that profiles are all defined within one QIR version, itself annotated with these module attributes. To recap another point from our discussion, where it will get perhaps unwieldy for developers is when a breaking change is made _to one profile_ but not to other profiles or the overall spec, or is made to a part of the spec not used by a particular profile. This will particularly be the case as new profiles are added but not yet stabilized - SomeNewProfile may cause the BaseProfile to see several new major versions without any changes of its own. This is where we may benefit from version numbers attached to the profile as well (not just QIR), but that can be added later and nothing in this PR excludes that possibility.

tl;dr: sounds good to me as-is, no changes requested 👍

Reply:
Thanks for the discussions. I agree that we will need to gain some experience value in how to establish new functionality/profiles - it is hard to have confidence in their specification without having a reasonably complete implementation, but it is also unfortunate if any "real world testing" of a first implementation is not formally covered by a spec with a version number. I think we may just need a "prerelease version" of the spec as well. I think we can adjust how we deal with that as we gain some experience here. 
As a note, I made some tweaks to the phrasing to reflect that the invocation of an entry point as part of a full application (where the full application would also include the service submission etc) is not captured in the profile itself. That gives that freedom to use this profile in combination with different service features without technically being incompliant with the spec.